OpenSim/Moco/MocoGoal/MocoImpulseTrackingGoal.h

```diff
@@ -0,0 +1,283 @@
+            : MocoGoal(std::move(name), weight) {
+            constructProperties();
+        }
+
+        /// Set the ExternalLoads as an XML file. This clears the ExternalLoads
+        /// provided as an object, if one exists.
+        void setExternalLoadsFile(const std::string& extLoadsFile);
+        /// Set the ExternalLoads as an object. This clears the ExternalLoads
+        /// XML file, if provided.
+        void setExternalLoads(const ExternalLoads& extLoads);
+
+        /// Add a group of contact forces whose sum should track the force data from
+        /// a single ExternalForce. The externalForceName should be the name of an
+        /// ExternalForce object in the ExternalLoads.
+        void addContactGroup(
+            const std::vector<std::string>& contactForcePaths,
+            const std::string& externalForceName) {
+            append_contact_groups(MocoImpulseTrackingGoalGroup(
+                contactForcePaths, externalForceName));
+        }
+        /// Add a group of contact forces whose sum should track the force data from
+        /// a single ExternalForce.
+        /// If the contact force elements associated with a single ExternalForce are
+        /// distributed across multiple bodies use this function instead of the
+        /// easier-to-use addContactGroup(), and set the group's
+        /// alternative_frame_paths property accordingly. See
+        /// MocoContactTrackingGoalGroup for more information.
+        void addContactGroup(MocoImpulseTrackingGoalGroup group) {
+            append_contact_groups(std::move(group));
+        }
+
+        /// Set the GRF Impulse 'axis' to be tracked. As per OpenSim convention:
+        /// X = 0; 
+        /// Y = 1; 
+        /// Z = 2;
+        void setImpulseAxis(int ImpulseAxis) {
+            set_impulse_axis(ImpulseAxis);
+        }
+        int getImpulseAxis() const { return get_impulse_axis(); }
+
+        /// Add a MocoParameter to the problem that will scale the tracking reference
+        /// data associated with a contact force group. Scale factors are applied
+        /// to the tracking error calculations based on the following equation:
+        ///
+        ///     error = modelValue - scaleFactor * referenceValue
+        ///
+        /// In other words, the scale factor is applied when computing the tracking
+        /// error for each contact force group, not to the reference data directly.
+        /// You must specify both the external force name associated with the contact
+        /// force group and the index corresponding to the direction (i.e., X = 0,
+        /// Y = 1, Z = 2) of the scaled force value. The direction is applied in
+        /// whatever frame the reference data is expressed in based on the provided
+        /// ExternalLoads in each contact group.
+        void addScaleFactor(const std::string& name,
+            const std::string& externalForceName, int index,
+            const MocoBounds& bounds);
```

Comment:
I have replaced `int index` with `int impulse_axis` ... but I am now wondering what impact this will have on 'mutable std::map<std::pair<std::string, int>, std::string> m_scaleFactorMap;
        using RefPtrMSF = SimTK::ReferencePtr<const MocoScaleFactor>;
        mutable std::vector<std::array<RefPtrMSF, 3>> m_scaleFactorRefs;'

Reply:
ignore - saw your comment below!